### PR TITLE
Add E2E tests for the integration in Product Block Editor

### PR DESCRIPTION
**Pull request by eason9487**
### Changes proposed in this Pull Request:

- Upgrade `@playwright/test` to `^1.42.0` for having the new APIs:
   - https://playwright.dev/docs/api/class-page#page-unroute-all
   - https://playwright.dev/docs/api/class-locator#locator-press-sequentially
- Convert space indentations in `test-snippets.php` to tabs.
- Adjust the product data in `tests/e2e/config/default.json` to fit with WC v3 REST API and remove unused parts.
- Add new E2E API utils for creating variable and variation products.
- Add E2E tests

#### 📌 Checklist (@eason9487)

- [x] Revert 77022c71d358d87a942112c13f9c36b6dca25b33 before merging this PR

### Detailed test instructions:

1. `npm install` to update `@playwright/test` package.
2. `npx playwright install chromium` to update the PlayWright browser as well.
3. `npm run dev` to build files
4. `npm run test:e2e` to see if the E2E test in headless mode can pass.
5. (Optional) `npm run test:e2e -- --ui` to inspect the E2E test in UI mode.
6. View the E2E test result on GitHub Actions: https://github.com/woocommerce/google-listings-and-ads/actions/runs/8376570038/job/22936559491?pr=2327

### CODE REVIEW

File: .github/workflows/e2e-tests.yml
Comment (puntope) [inline .github/workflows/e2e-tests.yml]: Does this mean we will run the e2e tests on each PR?

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js:27]: Why have this line here if we are clearing it after?

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js:44]: Based on the comment. Makes more sense for me to add this to the next step.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js:66]: I guess at this point we are already in ProductPage?  Why not just to change the tab and save a page load?

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js]: I wonder if might be better option to just check the link href and save a page load (for reducing time test time)

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js:59]: This is being tested I guess in line 54 as the default product is standard. So we can omit this

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js:287]: Why are we testing again this part? Maybe we can reduce execution time as 'Hide in product catalog' is false by default and is tested already.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js:292]: I guess these 2 can be omitted and save a page load

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js:378]: These 2 can be omitted and save a page load

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (puntope) [inline tests/e2e/specs/product-editor/block-integration.test.js:684]: Why we need this?

File: .github/workflows/e2e-tests.yml
Comment (puntope) [reply .github/workflows/e2e-tests.yml]: Ok, I see you add it for remove it

**Review by puntope — COMMENTED** [review]
Hi @eason9487 Thanks for your work in E2E test.

All is working fine. Not a strong blocker. However, I worry about the time for the tests to finish execution.

2-3mins

Screenshot 2024-03-24 at 12 31 15

I left some comments in an attempt to reduce unnecessary saves/loads. I will wait for your feedback before approve.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js:27]: Apart from the "Prompt to Get Started" case, all other cases need to be tested in the onboarded state.

If it isn't set here, then manual setup will be required when developing other cases, or the "Prompt to Get Started" case will have to be run first.

The main goal is to get the test cases as close to atomicity as possible so that test cases can be tested individually. This makes development and debugging easier and reduces false positives caused by side effects.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js:44]: It's related to https://github.com/woocommerce/google-listings-and-ads/pull/2327#discussion_r1537142428.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js:66]: This is also to make the test case as close as possible to atomicity.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js]: Yes, it can be reduced by about 1 ~ 2 seconds. Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2327/commits/caeefeeedff5bb47442cc9ffcb933be98935686b.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js:59]: This asserts that the tab can be changed from hidden to visible, which is not equivalent to being visible by default. Also, for the next one assert that the tab goes from visible to hidden.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js:287]: This asserts that the channel visibility status can be changed back from disabled to enabled.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js:292]: This is also to make the test case as close as possible to atomicity.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js:378]: This is also to make the test case as close as possible to atomicity.

File: tests/e2e/specs/product-editor/block-integration.test.js
Comment (eason9487) [reply tests/e2e/specs/product-editor/block-integration.test.js:684]: This is to removes all routes created with `page.route()` via the `editorUtils.mockChannelVisibility` calls in the "Change channel visibility and check its notice, status, and issues" case. In this way, the side effects between test cases can be reduced.

**Comment by eason9487** [comment]
Hi @puntope, thanks for the review. Could you take another look?

> However, I worry about the time for the tests to finish execution.

The long execution time is mainly because the Product Block Editor page loading takes 6~10 seconds, which is rather uncontrollable from the plugin side.

Currently, each test has at least one atomicity that can be executed independently and has no side effects. Allowing the next case to continue where the finished states of the previous one would reduce some overall time but at the cost of making development and debugging more difficult. I'm leaning toward keeping test cases as atomic as possible.

**Review by puntope — APPROVED** [review]
Thanks for the clarifications @eason9487 

Sad the Product Editor loads causes those delays. I'm approving it since it doesn't seem to have any action on our side to mitigate it.

**Pull request merged**
